### PR TITLE
feat: Add ODP GraphQL Manager and Tests

### PR DESCRIPTION
**Pull request by mikechu-optimizely**
## Summary
- Added an GraphQLManager interface and default implementation for querying qualified segments
- GraphQLManager is also responsible for parsing response from ODP. 

## Test plan
- Added specific GraphQL request and response test class

## Issues
- OASIS-8406

### CODE REVIEW

File: OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs
Comment (jaeopt) [inline OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs:259]: This case should be considered as an error (JSON parse error, or similar), logging error message and return an empty array (instead of null).

File: OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs
Comment (jaeopt) [inline OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs:263]: Any way to inject a network error?

File: OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs
Comment (jaeopt) [inline OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs:78]: Include a node with "not-qualified" state to test filtering.

File: OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs
Comment (jaeopt) [inline OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs:215]: Can we check if all error log messages include "Audience segments fetch failed (...)"?

**Review by jaeopt — CHANGES_REQUESTED** [review]
Looks good. Some suggestions for error handling.
I see some C# stuffs I cannot review. Let Sohail double check.

File: OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs
Comment (mikechu-optimizely) [reply OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs:263]: I did forget to develop this test. Apologies.

I looked Friday and again this morning. I can't seem to find a way to cause or simulate a network error within a test other than to inject a desired response status code.  As an alternative, I've restructured to handle 4XX and 5XX responses.

File: OptimizelySDK/Odp/Client/OdpClient.cs
Comment (jaeopt) [inline OptimizelySDK/Odp/Client/OdpClient.cs:50]:
```suggestion
                _logger.Log(LogLevel.ERROR, "Audience segments fetch failed (network error)");
```

File: OptimizelySDK/Odp/Client/OdpClient.cs
Comment (jaeopt) [inline OptimizelySDK/Odp/Client/OdpClient.cs:57]:
```suggestion
                _logger.Log(LogLevel.ERROR, "Audience segments fetch failed (<response.statusCode or statusMessage here>)");
```

**Review by jaeopt — CHANGES_REQUESTED** [review]
Changes look good. Some more suggestions regarding error handling.

**Review by jaeopt — APPROVED** [review]
LGTM

**Comment by mikechu-optimizely** [comment]
> LGTM

@jaeopt 

Should I wait until @msohailhussain reviews?

**Comment by jaeopt** [comment]
> > LGTM
> 
> @jaeopt
> 
> Should I wait until @msohailhussain reviews?

@mikechu-optimizely Yes, let's wait for his review.

**Comment by mikechu-optimizely** [comment]
Hopefully, I've not ruined your codebase with this PR @msohailhussain  😁

**Comment by mikechu-optimizely** [comment]
@msohailhussain when you have a chance, this PR also needs your 👍   Thanks!

File: OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs
Comment (msohailhussain) [inline OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs:109]: Please change this message, java.lang may confuse.

File: OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs
Comment (msohailhussain) [inline OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs:187]: java.lang.Runtime**

File: OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs
Comment (msohailhussain) [inline OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs:204]: Please add specific message, what we are expecting.

File: OptimizelySDK/Odp/Client/IOdpClient.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/Client/IOdpClient.cs:23]: add documentation

File: OptimizelySDK/Odp/Client/OdpClient.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/Client/OdpClient.cs:44]: logger=null

File: OptimizelySDK/Odp/Client/OdpClient.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/Client/OdpClient.cs:46]: `_logger = logger ?? new NoOpLogger()`

File: OptimizelySDK/Odp/Client/OdpClient.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/Client/OdpClient.cs:60]: can't we just call `QuerySegmentsAsync(parameters).Result`

File: OptimizelySDK/Odp/Client/OdpClient.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/Client/OdpClient.cs:64]: Can you please add ErrorHandler as well?

File: OptimizelySDK/Odp/Client/OdpClient.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/Client/OdpClient.cs:112]: take const.

File: OptimizelySDK/Odp/Client/OdpClient.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/Client/OdpClient.cs:115]: same as above.

File: OptimizelySDK/Odp/Entity/Data.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/Entity/Data.cs:19]: Document all `Customer`, `Data`, `Edge` classes.

File: OptimizelySDK/Odp/Entity/Extension.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/Entity/Extension.cs:19]: doc?

File: OptimizelySDK/Odp/Entity/QuerySegmentsParameters.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/Entity/QuerySegmentsParameters.cs:59]: Can we make a better implementation for `ToJson`

File: OptimizelySDK/Odp/Entity/Response.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/Entity/Response.cs:19]: doc

File: OptimizelySDK/Odp/GraphQLManager.cs
Comment (msohailhussain) [reply OptimizelySDK/Odp/GraphQLManager.cs:38]: Can you please check in our existing SDK, what we use in case of `null`, I think we use `NoOpLogger.

File: OptimizelySDK/Odp/GraphQLManager.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/GraphQLManager.cs:36]: document it. what kind of logger and OdpClient by default is assigned.

File: OptimizelySDK/Odp/GraphQLManager.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/GraphQLManager.cs:67]: what you are returning ? string.Empty should work if it's returning empty.

File: OptimizelySDK/Odp/GraphQLManager.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/GraphQLManager.cs:88]: we don't need `?`, already checked in above logic.

File: OptimizelySDK/Odp/IGraphQLManager.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/IGraphQLManager.cs:21]: doc

**Review by msohailhussain — CHANGES_REQUESTED** [review]
please address.

File: OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs
Comment (mikechu-optimizely) [reply OptimizelySDK.Tests/OdpTests/GraphQLManagerTest.cs:36]: I think it's a test key.
https://github.com/optimizely/swift-sdk/blob/84875a42622d1da89b5ca0284ba7a73b7e0ff55e/Sources/ODP/ZaiusGraphQLApiManager.swift#L21

File: OptimizelySDK/Odp/Client/OdpClient.cs
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/Client/OdpClient.cs:60]: As I understand it, `.GetAwaiter().GetResult()` propagates exceptions back to this thread. I believe this may be desirable.

File: OptimizelySDK/Odp/Entity/Audience.cs
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/Entity/Audience.cs:21]: `JsonConvert.DeserializeObject` fails when `private set;`

File: OptimizelySDK/Odp/Entity/Data.cs
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/Entity/Data.cs:22]: I came up with some pretty weak descriptions.  Sorry.  I just couldn't find any good Zaius data dictionary info.

@msohailhussain @jaeopt

File: OptimizelySDK/Odp/Client/OdpClient.cs
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/Client/OdpClient.cs:61]: Do you have any code suggestions?  I'm stuck.

File: OptimizelySDK/Odp/Entity/QuerySegmentsParameters.cs
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/Entity/QuerySegmentsParameters.cs]: I renamed this a bit to be more clear, but I'm not sure what improvements we could make to produce a GraphQL JSON string.  

Got any suggestions?

File: OptimizelySDK/Odp/Client/OdpClient.cs
Comment (msohailhussain) [reply OptimizelySDK/Odp/Client/OdpClient.cs:66]: https://github.com/optimizely/csharp-sdk/blob/master/OptimizelySDK/Config/HttpProjectConfigManager.cs#L317

File: OptimizelySDK/Odp/Client/OdpClient.cs
Comment (msohailhussain) [reply OptimizelySDK/Odp/Client/OdpClient.cs:62]: I will discuss offline.

File: OptimizelySDK/Odp/Entity/QuerySegmentsParameters.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/Entity/QuerySegmentsParameters.cs:31]: great.

**Review by msohailhussain — APPROVED** [review]
before merge, please discuss GetAwaiter with me.

**Pull request merged**
